src/mscorlib/shared/System/Threading/ReaderWriterLockSlim.cs

```diff
@@ -1059,33 +1121,273 @@ private uint GetNumReaders()
             return _owners & READER_MASK;
         }
 
+        private bool ShouldSpinForEnterAnyRead()
+        {
+            // If there is a write waiter or write upgrade waiter, the waiter would block a reader from acquiring the RW lock
+            // because the waiter takes precedence. In that case, the reader is not likely to make progress by spinning.
+            return _fNoWaiters || _numWriteWaiters == 0 && _numWriteUpgradeWaiters == 0;
+        }
+
+        private bool ShouldSpinForEnterAnyWrite(bool isUpgradeToWrite)
+        {
+            // If there is a write upgrade waiter, the waiter would block a writer from acquiring the RW lock because the waiter
+            // holds a read lock. In that case, the writer is not likely to make progress by spinning.
+            return isUpgradeToWrite || _numWriteUpgradeWaiters == 0;
```

Comment:
There can only be one upgradeable read lock held at a time, so if this is a write upgrade, there can't be any write upgrade waiters.